WordPress/Classes/Services/ReaderPostService.m

```diff
@@ -589,10 +589,6 @@ - (void)flagPostsFromSite:(NSNumber *)siteID asBlocked:(BOOL)blocked
     for (ReaderPost *post in results) {
         post.isSiteBlocked = blocked;
     }
-
-    [self.managedObjectContext performBlockAndWait:^{
-        [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
-    }];
```

Comment:
This call to save the changes to the context has been removed because the consumer code now calls this method within a `performAndSave` (right?)

I can see a pattern emerging here, where we have functions that take a context as input and modify it but delegate persisting the changes to their callers. Is that a "safe" pattern, safe as in a pattern that other developers will adopt by default?

I guess the worst that can happen would be for folks writing/updating methods such as this to add an explicit call to save, which would only (?) end up in wasted computation?

---

**Update** In a later commit, this method is moved in the same scope/file as the caller. In such case, my question is less important because this now standalone code works more as a DRY tool than as a method accessible to the outside.

Reply:
That's a really good question. One of many things I find working with Core Data quite hard is, you almost _must_ have some kind of convention[^1] when dealing with `NSManagedObjectContext` and `NSManagedObject`, due to Core Data's nature of not being thread-safe. Like the pattern you pointed out here, we either
1. make the function saves the `NSManagedObjectContext`, which means the caller don't need to worry about if the changes are saved or not, but all other functions that make changes to a context object must do the same. Then, we end up having many function making small changes to the database. Or,
1. delegate the saving to the call site, so that only one saving is needed (which seems good since there is less overhead). That means, we'll need to have some kind of convention to communicate the idea of "this function only makes changes, doesn't save the context".

Similar problem exists with using `NSManagedObject` instances. Since it's only safe to use the instance within it's context queue, when we have functions that uses `NSManagedObject`, how can we guarantee its thread-safey? Do we
1. make the functions internally call `aManagedObject.managedContext.perform()` before accessing the objects' properties, which seems paranoid and I imagine there will be much overhead. Or,
2. delegate the responsibility of calling the function in the right thread to the caller, which doesn't seem safe at all.

[^1]: By "convention", I mean something can't be enforced syntactically by using Swift language feature and must be agreed upon among all developers.

Reply:
Interesting... 🤔 

I think it's worth establishing and documenting conventions and their rationales.

Since both convention options rely on us trusting our future selves to remember to do something, I'd go for the one that results in cleaner code. I think at this point it would be option 2 in both cases.